### DISCUSSION TITLE
Accuracy not improving on NMNIST dataset

### DISCUSSION THREAD

**Discussion by muhammedgouda** [jeshraghian/snntorch#139] [Q&A]
Hi all, 

I have tried to run the notebook of tutorial 7 myself. First I got this error in the training loop 
![image](https://user-images.githubusercontent.com/72167749/199011272-09883c78-1a8a-492c-bcc9-a1d426a8de08.png)

I got rid of this error by reshaping the spk_rec tensor like this:

net.train()
spk_rec = forward_pass(net, data)
**spk_rec = torch.reshape(spk_rec, (spk_rec.size()[1],spk_rec.size()[0],spk_rec.size()[2]))**
loss_val = loss_fn(spk_rec, targets)

However, after running the training loop the accuracy didn't seem to be improving as you can see in the following plot. 

![image](https://user-images.githubusercontent.com/72167749/199013009-ab9157df-8f08-4b4c-84d2-41775366788a.png)

I'm wondering if my reshaping line caused other problems. Does anyone have an idea ? 

I also plotted the train loss and it looks like this 

![image](https://user-images.githubusercontent.com/72167749/199014061-4b534d54-6128-489f-81bc-40c08647a318.png)

**Comment by ridgerchu** [comment]
Hi,
 
I checked your code and found that you used the PyTorch default MSE loss. However, if you need to employ it, you need to encoder the label with one-hot. I suggest you could try the snnTorch’s MSE Loss, which can be found in `snntorch.functional.loss.mse_count_loss` and its doc can be found in [here](https://snntorch.readthedocs.io/en/latest/snntorch.functional.html#snntorch.functional.loss.mse_count_loss)

**Reply by muhammedgouda** [reply]
Hi, 

Thanks for your reply. In my code I'm already using the mse loss function from snntorch 
loss_fn = SF.mse_count_loss(correct_rate=0.8, incorrect_rate=0.2)

**Reply by ridgerchu** [reply]
Hi,

Could you please post your full network training code under this issue? it will help me to find out the error.

**Reply by muhammedgouda** [reply]
Hi, 

Yes, here is my code:

![image](https://user-images.githubusercontent.com/72167749/199373157-746d7c2a-5c58-4a99-9db7-4d4cf35be068.png)

**Reply by ridgerchu** [reply]
Hi,

I've not found any problems in this code. But I'm weird about the tensor dimension of the first error (torch.Size[309,10]), it may indicate that you change the network structure. Do you just run the jupyter notebook of tutorial 7 without any modification?

**Reply by muhammedgouda** [reply]
Hi, 

Yes I didn't modify anything else in the code. 

But there is something I just found out now. It turns out if I don't invert the shape of the spike tensor inside the training loop, and rather invert the data tensor in the forward_pass() method, the training works fine. 
![image](https://user-images.githubusercontent.com/72167749/199394037-b804a469-a3c9-4b11-9156-c32045456c43.png)


Here is a screen shot of the accuracy plot 
![image](https://user-images.githubusercontent.com/72167749/199394120-e032be40-47c5-43ed-9b55-2a02d5249e65.png)


So this seems to solve my problem. I still don't know why inverting the spike tensor caused that issue in the progression of the accuracy but not in the loss though.

**Reply by jeshraghian** [reply]
I think Tonic might have made a few changes to the tensor shape since I wrote this tutorial. We should definitely fix this in both the notebook and in the static html docs. Thanks for flagging this!

**Reply by ridgerchu** [reply]
Hi,
 
Thanks for posting the solution! I hope it will function well~
